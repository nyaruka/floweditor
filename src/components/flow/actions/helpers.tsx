import {
  Contact,
  Endpoints,
  Group,
  RecipientsAction,
  FlowIssue,
  FlowIssueType,
  BroadcastMsg,
  SendMsg
} from 'flowTypes';
import * as React from 'react';
import { Asset, AssetType } from 'store/flowContext';
import { FormEntry, NodeEditorSettings, ValidationFailure } from 'store/nodeEditor';
import { createUUID } from 'utils';
import { Trans } from 'react-i18next';
import shared from 'components/shared.module.scss';
import { showHelpArticle } from 'external';
import { IssueProps } from '../props';

export const renderIssues = (issueProps: IssueProps): JSX.Element => {
  const { issues, helpArticles } = issueProps;
  if (!issues || issues.length === 0) {
    return null;
  }

  return (
    <div style={{ padding: '10px 0px' }}>
      {issues.map((issue: FlowIssue, num: Number) => {
        const key = issue.node_uuid + issue.action_uuid + num;
        return (
          <div
            style={{ margin: '6px 0px', display: 'flex', fontSize: '14px', color: 'tomato' }}
            key={key}
          >
            <div
              style={{ marginRight: '8px', marginTop: '-2px', fontSize: '18px' }}
              className={`fe-warning`}
            />
            <div>{renderIssue(issue, helpArticles)}</div>
          </div>
        );
      })}
    </div>
  );
};

export const renderIssue = (
  issue: FlowIssue,
  helpArticles: { [key: string]: string } = {}
): JSX.Element => {
  // worst case, defer to the default description
  let message: JSX.Element = <>{issue.description}</>;

  if (issue.type === FlowIssueType.MISSING_DEPENDENCY) {
    message = (
      <Trans
        i18nKey="issues.missing_dependency"
        values={{
          name: issue.dependency.name || issue.dependency.key,
          type: issue.dependency.type
        }}
      >
        Cannot find a [[type]] for <span className="emphasize">[[name]]</span>
      </Trans>
    );
  }

  if (issue.type === FlowIssueType.INVALID_REGEX) {
    message = (
      <Trans i18nKey="issues.legacy_extra" values={{ regex: issue.regex }}>
        Invalid regular expression found: [[regex]]
      </Trans>
    );
  }

  if (issue.type === FlowIssueType.LEGACY_EXTRA) {
    message = (
      <Trans i18nKey="issues.legacy_extra">Expressions should not reference @legacy_extra</Trans>
    );
  }

  const article = helpArticles[issue.type];
  if (article) {
    return (
      <div
        className={shared.issue_help}
        onClick={() => {
          showHelpArticle(article);
        }}
      >
        {message}
      </div>
    );
  }

  return message;
};

export const getActionUUID = (nodeSettings: NodeEditorSettings, currentType: string): string => {
  if (nodeSettings.originalAction && nodeSettings.originalAction.type === currentType) {
    return nodeSettings.originalAction.uuid;
  }
  return createUUID();
};

export const getCompose = (action: SendMsg | BroadcastMsg = null): string => {
  if (!action) {
    return getEmptyComposeValue();
  }
  return action.compose;
};

export const getEmptyComposeValue = (): string => {
  return JSON.stringify({ text: '', attachments: [] });
};

export const getRecipients = (action: RecipientsAction): Asset[] => {
  let selected: any[] = (action.groups || []).map((group: Group) => {
    return {
      id: group.uuid,
      name: group.name,
      type: AssetType.Group
    };
  });

  selected = selected.concat(
    (action.contacts || []).map((contact: Contact) => {
      return { id: contact.uuid, name: contact.name, type: AssetType.Contact, missing: false };
    })
  );

  selected = selected.concat(
    (action.legacy_vars || []).map((expression: string) => {
      return { name: expression, value: expression, expression: true };
    })
  );

  return selected;
};

export const renderAssetList = (
  assets: Asset[],
  max: number = 10,
  endpoints: Endpoints
): JSX.Element[] => {
  // show our missing ones first
  return assets.reduce((elements, asset, idx) => {
    if (idx <= max - 2 || assets.length === max) {
      elements.push(renderAsset(asset, endpoints));
    } else if (idx === max - 1) {
      elements.push(<div key="ellipses">+{assets.length - max + 1} more</div>);
    }
    return elements;
  }, []);
};

export const renderAsset = (asset: Asset, endpoints: Endpoints) => {
  let assetBody = null;

  switch (asset.type) {
    case AssetType.Classifier:
      assetBody = (
        <Trans i18nKey="assets.classifier" values={{ name: asset.name }}>
          Call [[name]] classifier
        </Trans>
      );
      break;
    case AssetType.Group:
      assetBody = (
        <>
          <span className={`${shared.node_group} fe-group`} />
          {asset.name}
        </>
      );
      break;
    case AssetType.Label:
      assetBody = (
        <>
          <span className={`${shared.node_label} fe-label`} />
          {asset.name}
        </>
      );
      break;
    case AssetType.Flow:
      assetBody = (
        <>
          <span className={`${shared.node_label} fe-split`} />
          <a
            onMouseDown={(e: any) => {
              e.preventDefault();
              e.stopPropagation();
            }}
            onMouseUp={(e: any) => {
              e.preventDefault();
              e.stopPropagation();
            }}
            href={`${endpoints.editor}/${asset.id}`}
            rel="noopener noreferrer"
            target="_blank"
          >
            {asset.name}
          </a>
        </>
      );
      break;
    case AssetType.Ticketer:
      assetBody = (
        <Trans i18nKey="assets.ticketer" values={{ name: asset.name }}>
          Open a new Ticket on [[name]]
        </Trans>
      );
      break;
  }

  if (!assetBody) {
    assetBody = asset.name;
  }

  return (
    <div className={`${shared.node_asset}`} key={asset.id || (asset as any).value}>
      {assetBody}
    </div>
  );
};

export const getAllErrors = (entry: FormEntry): ValidationFailure[] => {
  return entry ? entry.validationFailures || [] : [];
};

export const hasErrors = (entry: FormEntry): boolean => {
  return getAllErrors(entry).length > 0;
};

export const getAllErrorMessages = (entry: FormEntry): string[] => {
  const errors = getAllErrors(entry).map((failure: ValidationFailure) => failure.message);
  return errors;
};

export const getExpressions = (assets: any[]): any[] => {
  return assets
    .filter((asset: any) => asset.expression)
    .map((asset: any) => {
      return asset.value;
    });
};

export const getRecipientsByAsset = (assets: Asset[], type: AssetType): any[] => {
  return assets
    .filter((asset: Asset) => asset.type === type)
    .map((asset: Asset) => {
      return { uuid: asset.id, name: asset.name };
    });
};

export const getComposeByAsset = (value: string, asset: string): any | any[] => {
  return JSON.parse(value)[asset];
};
