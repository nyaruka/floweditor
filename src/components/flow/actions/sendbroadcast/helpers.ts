import {
  getActionUUID,
  getComposeActionToState,
  getComposeStateToAction,
  getRecipients,
  getRecipientsStateToAction
} from 'components/flow/actions/helpers';
import { SendBroadcastFormState } from 'components/flow/actions/sendbroadcast/SendBroadcastForm';
import { Types } from 'config/interfaces';
import { BroadcastMsg } from 'flowTypes';
import { NodeEditorSettings } from 'store/nodeEditor';

export const initializeForm = (settings: NodeEditorSettings): SendBroadcastFormState => {
  if (settings.originalAction && settings.originalAction.type === Types.send_broadcast) {
    let action = settings.originalAction as BroadcastMsg;

    // check if our form should use a localized action
    if (settings.localizations && settings.localizations.length > 0) {
      const localized = settings.localizations[0];
      if (localized.isLocalized()) {
        action = settings.localizations[0].getObject() as BroadcastMsg;
      } else {
        return {
          compose: { value: getComposeActionToState() },
          recipients: { value: [] },
          valid: true
        };
      }
    }

    return {
      compose: { value: getComposeActionToState(action) },
      recipients: { value: getRecipients(action) },
      valid: true
    };
  }

  return {
    compose: { value: getComposeActionToState() },
    recipients: { value: [] },
    valid: false
  };
};

export const stateToAction = (
  settings: NodeEditorSettings,
  formState: SendBroadcastFormState
): BroadcastMsg => {
  const [compose, text, attachments] = getComposeStateToAction(formState);
  const [legacy_vars, contacts, groups] = getRecipientsStateToAction(formState);

  return {
    legacy_vars: legacy_vars,
    contacts: contacts,
    groups: groups,
    compose: compose,
    text: text,
    attachments: attachments,
    type: Types.send_broadcast,
    uuid: getActionUUID(settings, Types.send_broadcast)
  };
};
