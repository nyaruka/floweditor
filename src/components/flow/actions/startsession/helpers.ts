import {
  getActionUUID,
  getExpressions,
  getRecipients,
  getRecipientsByAsset
} from 'components/flow/actions/helpers';
import {
  StartSessionFormState,
  START_TYPE_CREATE,
  START_TYPE_ASSETS,
  START_TYPE_QUERY
} from 'components/flow/actions/startsession/StartSessionForm';
import { Types } from 'config/interfaces';
import { StartSession } from 'flowTypes';
import { AssetType } from 'store/flowContext';
import { NodeEditorSettings } from 'store/nodeEditor';

export const initializeForm = (settings: NodeEditorSettings): StartSessionFormState => {
  if (settings.originalAction && settings.originalAction.type === Types.start_session) {
    const action = settings.originalAction as StartSession;

    const init = {
      recipients: {
        value: getRecipients(action)
      },
      flow: {
        value: action.flow
      },
      startType: {
        value: action.create_contact
          ? START_TYPE_CREATE
          : action.contact_query
          ? START_TYPE_QUERY
          : START_TYPE_ASSETS
      },
      contactQuery: { value: action.contact_query || '' },
      exclusions: action.exclusions,
      valid: true
    };

    return init;
  }

  return {
    recipients: { value: [] },
    flow: { value: null },
    startType: { value: START_TYPE_ASSETS },
    contactQuery: { value: '' },
    exclusions: { in_a_flow: false },
    valid: false
  };
};

export const stateToAction = (
  settings: NodeEditorSettings,
  state: StartSessionFormState
): StartSession => {
  const flow = state.flow.value;

  const action: StartSession = {
    legacy_vars: getExpressions(state.recipients.value),
    contacts: getRecipientsByAsset(state.recipients.value, AssetType.Contact),
    groups: getRecipientsByAsset(state.recipients.value, AssetType.Group),
    create_contact: state.startType.value === START_TYPE_CREATE,
    flow: { name: flow.name, uuid: flow.uuid },
    type: Types.start_session,
    uuid: getActionUUID(settings, Types.start_session),
    exclusions: state.exclusions
  };

  // only include contact query if it is set
  if (state.contactQuery.value) {
    action['contact_query'] = state.contactQuery.value;
  }

  return action;
};
