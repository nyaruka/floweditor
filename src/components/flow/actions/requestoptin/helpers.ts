import { getActionUUID } from 'components/flow/actions/helpers';
import { Types } from 'config/interfaces';
import { RequestOptIn } from 'flowTypes';
import { NodeEditorSettings } from 'store/nodeEditor';
import { RequestOptInFormState } from './RequestOptInForm';

export const initializeForm = (settings: NodeEditorSettings): RequestOptInFormState => {
  if (settings.originalAction && settings.originalAction.type === Types.request_optin) {
    const action = settings.originalAction as RequestOptIn;
    return {
      optin: { value: action.optin },
      valid: true
    };
  }

  return {
    optin: { value: null },
    valid: false
  };
};

export const stateToAction = (
  settings: NodeEditorSettings,
  formState: RequestOptInFormState
): RequestOptIn => {
  const result = {
    type: Types.request_optin,
    optin: formState.optin.value[0],
    uuid: getActionUUID(settings, Types.add_input_labels)
  };
  return result;
};
