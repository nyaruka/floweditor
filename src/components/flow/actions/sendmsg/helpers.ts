/* eslint-disable @typescript-eslint/explicit-function-return-type */
import { getActionUUID } from 'components/flow/actions/helpers';
import { SendMsgFormState } from 'components/flow/actions/sendmsg/SendMsgForm';
import { Types } from 'config/interfaces';
import { FlowEditorConfig, Label, MsgTemplating, SendMsg } from 'flowTypes';
import { FormEntry, NodeEditorSettings, StringEntry } from 'store/nodeEditor';
import { SelectOption } from 'components/form/select/SelectElement';
import { createUUID } from 'utils';
import { Attachment } from './attachments';

export const TOPIC_OPTIONS: SelectOption[] = [
  { value: 'event', name: 'Event' },
  { value: 'account', name: 'Account' },
  { value: 'purchase', name: 'Purchase' },
  { value: 'agent', name: 'Agent' }
];

export const initializeForm = (
  settings: NodeEditorSettings,
  config: FlowEditorConfig
): SendMsgFormState => {
  let template: FormEntry = { value: null };
  let templateVariables: StringEntry[] = [];
  let skipValidation = false;

  if (settings.originalAction && settings.originalAction.type === Types.send_msg) {
    const action = settings.originalAction as SendMsg;
    const attachments: Attachment[] = [];
    let expressionValue = null;
    (action.attachments || []).forEach((attachmentString: string) => {
      const splitPoint = attachmentString.indexOf(':');

      const type = attachmentString.substring(0, splitPoint);
      const attachment = {
        type,
        url: attachmentString.substring(splitPoint + 1),
        uploaded: type.indexOf('/') > -1,
        valid: false
      };

      attachments.push(attachment);
    });

    if (action.templating) {
      const msgTemplate = action.templating.template;
      if (action.templating.expression) {
        expressionValue = { value: action.templating.expression };
      }
      template = {
        value: {
          uuid: msgTemplate.uuid,
          name: msgTemplate.name
        }
      };
      templateVariables = action.templating.variables.map((value: string) => {
        return {
          value
        };
      });
    }

    const labels = action.labels
      ? action.labels.map((label: Label) => {
          if (label.name_match) {
            return { name: label.name_match, expression: true };
          }
          return label;
        })
      : [];

    if (config.skipValidation) {
      skipValidation = config.skipValidation;
    }

    return {
      expression: expressionValue,
      topic: { value: TOPIC_OPTIONS.find(option => option.value === action.topic) },
      template,
      templateVariables,
      attachments,
      labels: {
        value: labels
      },
      uploadInProgress: false,
      uploadError: '',
      message: { value: action.text },
      quickReplies: { value: action.quick_replies || [] },
      quickReplyEntry: { value: '' },
      sendAll: action.all_urns,
      valid: true,
      skipValidation
    };
  }

  return {
    topic: { value: null },
    template,
    templateVariables: [],
    attachments: [],
    uploadInProgress: false,
    uploadError: '',
    message: { value: '' },
    quickReplies: { value: [] },
    quickReplyEntry: { value: '' },
    sendAll: false,
    valid: false,
    labels: { value: [] }
  };
};

export const stateToAction = (settings: NodeEditorSettings, state: SendMsgFormState): SendMsg => {
  const attachments = state.attachments
    .filter((attachment: Attachment) => attachment.url.trim().length > 0)
    .map((attachment: Attachment) => `${attachment.type}:${attachment.url}`);

  let templating: MsgTemplating = null;

  if (state.template && state.template.value) {
    let templatingUUID = createUUID();
    if (settings.originalAction && settings.originalAction.type === Types.send_msg) {
      const action = settings.originalAction as SendMsg;
      if (
        action.templating &&
        action.templating.template &&
        action.templating.template.uuid === state.template.value.id
      ) {
        templatingUUID = action.templating.uuid;
      }
    }

    templating = {
      uuid: templatingUUID,
      template: {
        uuid: state.template.value.uuid,
        name: state.template.value.name
      },
      variables: state.templateVariables.map((variable: StringEntry) => variable.value)
    };

    if (state.expression && state.expression.value) {
      templating.expression = state.expression.value;
    }
  }
  const result: SendMsg = {
    attachments,
    text: state.message.value,
    type: Types.send_msg,
    all_urns: state.sendAll,
    labels: state.labels.value.map((label: any) => {
      if (label.expression) {
        return { name_match: label.name };
      }
      return label;
    }),
    quick_replies: state.quickReplies.value,
    uuid: getActionUUID(settings, Types.send_msg)
  };

  if (templating) {
    result.templating = templating;
  }

  if (state.topic.value) {
    result.topic = state.topic.value.value;
  }

  return result;
};
