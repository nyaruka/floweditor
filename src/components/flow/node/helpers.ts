import { getSwitchRouter } from 'components/flow/routers/helpers';
import { Types } from 'config/interfaces';
import {
  CallResthook,
  CallWebhook,
  Category,
  Exit,
  FlowNode,
  RouterTypes,
  TransferAirtime,
  Action,
  AnyAction,
  FlowIssue
} from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { getType } from 'config/typeConfigs';

export const getCategoriesForExit = (renderNode: RenderNode, exit: Exit): Category[] => {
  // if we are
  const isGroupSplit = getType(renderNode) === Types.split_by_groups;

  if (!renderNode.node.router) {
    return [];
  }
  return renderNode.node.router.categories
    .filter((cat: Category) => cat.exit_uuid === exit.uuid)
    .map((cat: Category) => {
      if (isGroupSplit) {
        return {
          ...cat
        };
      } else {
        return { ...cat, missing: false };
      }
    });
};

export const getResultName = (node: FlowNode) => {
  const switchRouter = getSwitchRouter(node);

  if (switchRouter && switchRouter.result_name) {
    return switchRouter.result_name;
  }

  if (node.router && node.router.type === RouterTypes.random) {
    return node.router.result_name;
  }

  if (node.actions.length === 1) {
    const action = node.actions[0];
    if (
      action.type === Types.call_webhook ||
      action.type === Types.call_resthook ||
      action.type === Types.open_ticket ||
      action.type === Types.transfer_airtime ||
      action.type === Types.link_google_sheet
    ) {
      const resultAction = action as CallWebhook | CallResthook | TransferAirtime;
      return resultAction.result_name;
    }
  }
};

export const getVisibleActions = (renderNode: RenderNode): Action[] => {
  // subflow nodes hide their set run results
  if (getType(renderNode) === Types.split_by_subflow) {
    return renderNode.node.actions.filter((action: Action) => action.type !== Types.set_run_result);
  }

  return renderNode.node.actions;
};

export const filterIssuesForAction = (
  nodeUUID: string,
  action: AnyAction,
  issues: FlowIssue[]
): FlowIssue[] => {
  return issues.filter(issue => issue.node_uuid === nodeUUID && issue.action_uuid === action.uuid);
};
