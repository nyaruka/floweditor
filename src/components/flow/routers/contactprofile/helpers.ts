import { getActionUUID } from 'components/flow/actions/helpers';
import { createWebhookBasedNode } from 'components/flow/routers/helpers';

import { Types } from 'config/interfaces';
import { SetContactProfile } from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings } from 'store/nodeEditor';

import { ContactProfileRouterFormState, profileOptions } from './ContactProfileRouterForm';

export const nodeToState = (settings: NodeEditorSettings): ContactProfileRouterFormState => {
  let resulNode: ContactProfileRouterFormState = {
    valid: true,
    optionType: { value: profileOptions['1'] },
    profileName: { value: '' },
    profileType: { value: '' }
  };

  if (settings.originalAction && settings.originalAction.type === Types.set_contact_profile) {
    const action = settings.originalAction as SetContactProfile;

    if (action.profile_type) {
      const option =
        Object.values(profileOptions).find(value => value.name === action.profile_type) ||
        profileOptions['1'];

      resulNode.optionType = {
        value: option
      };

      if (typeof action.value === 'string') {
        resulNode.profileName = { value: action.value };
      } else {
        resulNode.profileName = { value: action.value.name };
        resulNode.profileType = { value: action.value.type };
      }
    }
  }

  return resulNode;
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: ContactProfileRouterFormState
): RenderNode => {
  const { optionType, profileName, profileType } = state;

  const newAction: SetContactProfile = {
    profile_type: optionType.value.name,
    result_name: profileName.value,
    value:
      optionType.value.name === 'Create Profile'
        ? { name: profileName.value, type: profileType.value }
        : profileName.value,
    type: Types.set_contact_profile,
    uuid: getActionUUID(settings, Types.set_contact_profile)
  };

  return createWebhookBasedNode(newAction, settings.originalNode, false);
};
