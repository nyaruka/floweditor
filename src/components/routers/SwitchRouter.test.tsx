import * as React from 'react';
import { shallow, mount, ReactWrapper } from 'enzyme';
import { getSpecWrapper } from '../../helpers/utils';
import Config from '../../providers/ConfigProvider/configContext';
import SwitchRouterForm, {
    getListStyle,
    getItemStyle,
    composeExitMap,
    resolveExits,
    isSwitchRouterNode,
    SwitchRouterFormProps,
    SwitchRouterState
} from './SwitchRouter';
import CompMap from '../../services/ComponentMap';
import { LocalizedObject } from '../../services/Localization';
import { Exit, Case, SwitchRouter } from '../../flowTypes';
import { getLocalizations } from '../Node';
import NodeEditor from '../NodeEditor/NodeEditor';

const colorsFlow = require('../../../test_flows/a4f64f1b-85bc-477e-b706-de313a022979.json');
const formStyles = require('../NodeEditor/NodeEditor.scss');

const { baseLanguage, languages, getTypeConfig, getOperatorConfig, operatorConfigList } = Config;

const { results: [{ definition }] } = colorsFlow;
const { nodes: [replyNode, switchNodeMsg], localization: locals } = definition;

describe('SwitchRouter >', () => {
    describe('style utils >', () => {
        describe('getListStyle >', () => {
            it('should return "pointer" cursor style when passed a falsy isDraggingOver arg', () =>
                expect(getListStyle(false)).toEqual({
                    cursor: 'pointer'
                }));

            it('should return "move" cursor style when passed a truthy isDraggingOver arg', () =>
                expect(getListStyle(true)).toEqual({
                    cursor: 'move'
                }));
        });

        describe('getItemStyle >', () => {
            const notDraggingStyle = {
                transition: null,
                transform: null,
                pointerEvents: 'auto',
                WebkitTouchCallout: 'none',
                WebkitTapHighlightColor: 'rgba(0,0,0,0)',
                touchAction: 'manipulation'
            };

            const draggingStyle = {
                position: 'fixed',
                boxSizing: 'border-box',
                pointerEvents: 'none',
                zIndex: 5000,
                width: 595,
                height: 28,
                top: 271.3333435058594,
                left: 318.66668701171875,
                margin: 0,
                transform: 'translate(-2px, 17px)',
                WebkitTouchCallout: 'none',
                WebkitTapHighlightColor: 'rgba(0,0,0,0)',
                touchAction: 'manipulation'
            };

            it('should return not-dragging style when passed a falsy isDragging arg (snapshot)', () =>
                expect(getItemStyle(notDraggingStyle, false)).toMatchSnapshot());

            it('should return dragging style when passed a truthy isDragging arg (snapshot)', () =>
                expect(getItemStyle(draggingStyle, true)).toMatchSnapshot());
        });
    });

    describe('helpers >', () => {
        describe('resolveExits >', () =>
            it('should resolve exits', () => {
                const newCases = [
                    {
                        kase: {
                            uuid: '87173eee-5270-4233-aede-ca88e14b672a',
                            type: 'has_any_word',
                            exit_uuid: '7b245d49-e9e3-4387-b4ad-48deb03528cd',
                            arguments: ['red, r']
                        },
                        exitName: 'Red'
                    }
                ];

                const node = {
                    uuid: 'bc978e00-2f3d-41f2-87c1-26b3f14e5925',
                    router: {
                        type: 'switch',
                        default_exit_uuid: 'a8bdc1c5-0283-4656-b932-4f4094f4cc7e',
                        cases: [
                            {
                                uuid: '87173eee-5270-4233-aede-ca88e14b672a',
                                type: 'has_any_word',
                                exit_uuid: '7b245d49-e9e3-4387-b4ad-48deb03528cd',
                                arguments: ['red, r']
                            }
                        ],
                        operand: '@run.results.color '
                    },
                    exits: [
                        {
                            name: 'Red',
                            uuid: '7b245d49-e9e3-4387-b4ad-48deb03528cd',
                            destination_node_uuid: 'e2ecc8de-9774-4b74-a0dc-ca8aea123227'
                        },
                        {
                            uuid: 'a8bdc1c5-0283-4656-b932-4f4094f4cc7e',
                            name: 'Other',
                            destination_node_uuid: '533b64e2-5906-4d33-a8e9-64f1cb6c20dd'
                        }
                    ],
                    wait: {
                        type: 'exp'
                    }
                };

                expect(resolveExits(newCases, node)).toMatchSnapshot();
            }));

        describe('composeExitMap >', () =>
            it('should compose a map of exits', () =>
                expect(
                    composeExitMap([
                        {
                            uuid: '63e20770-64c1-4bad-bd73-a34cf6d46866',
                            name: 'All Responses',
                            destination_node_uuid: null
                        }
                    ])
                ).toMatchSnapshot()));

        describe('isSwitchRouterNode >', () => {
            it('should return true if node has a switch router', () =>
                expect(isSwitchRouterNode(switchNodeMsg)).toBeTruthy());

            it("should return false if node doesn't have a switch router", () =>
                expect(isSwitchRouterNode(replyNode)).toBeFalsy());
        });
    });

    describe('render >', () => {
        const config = getTypeConfig('wait_for_response');
        const ComponentMap = new CompMap(definition);
        const iso = 'spa';
        const translations = locals[iso];
        const localizations = getLocalizations(switchNodeMsg, iso, languages, translations);

        const nodeEditorContext = {
            getTypeConfig
        };

        const switchRouterContext = {
            getOperatorConfig,
            operatorConfigList
        };

        const spanish = { name: 'Spanish', iso: 'spa' };

        const nodeProps = {
            show: true,
            node: switchNodeMsg,
            language: spanish,
            definition,
            localizations,
            ComponentMap
        };

        const {
            onBindWidget,
            onBindAdvancedWidget,
            removeWidget,
            getExitTranslations,
            getLocalizedExits
        } = shallow(<NodeEditor {...nodeProps as any} />, {
            context: nodeEditorContext
        }).instance() as any;

        const switchProps = {
            node: switchNodeMsg,
            config,
            definition,
            ComponentMap,
            updateRouter: jest.fn(),
            onBindWidget,
            onBindAdvancedWidget,
            removeWidget,
            language: baseLanguage,
            showAdvanced: false,
            translating: false
        };

        const switchPropsTranslating = {
            ...switchProps,
            translating: true,
            language: spanish,
            localizations,
            getExitTranslations,
            getLocalizedExits
        };

        const placeholder: string = `${localizations[0].getLanguage().name} Translation`;

        it('should render wait_for_response form (not translating)', () => {
            // Cases
            const wrapper: ReactWrapper = mount(<SwitchRouterForm {...switchProps as any} />, {
                context: switchRouterContext
            });

            expect(getSpecWrapper(wrapper, 'case').length).toBe(8);

            expect(getSpecWrapper(wrapper, 'case-draggable').length).toBe(7);

            expect(
                getSpecWrapper(wrapper, 'case')
                    .last()
                    .prop('empty')
            ).toBeTruthy();

            /** Fields */
            expect(getSpecWrapper(wrapper, 'name-field').name()).toBe('TextInputElement');

            expect(getSpecWrapper(wrapper, 'name-field').props()).toEqual(
                expect.objectContaining({
                    'data-spec': 'name-field',
                    name: 'Result Name',
                    showLabel: true,
                    value: switchProps.node.router.result_name,
                    helpText:
                        'By naming the result, you can reference it later using @run.results.whatever_the_name_is',
                    ComponentMap: switchProps.ComponentMap
                })
            );

            expect(getSpecWrapper(wrapper, 'lead-in').text()).toBe('If the message response...');
        });

        it('should render wait_for_response form (translating)', () => {
            const wrapper: ReactWrapper = mount(
                <SwitchRouterForm {...switchPropsTranslating as any} />,
                {
                    context: switchRouterContext
                }
            );

            expect(getSpecWrapper(wrapper, 'title').exists()).toBeTruthy();

            expect(getSpecWrapper(wrapper, 'instructions').text()).toBe(
                'When category names are referenced later in the flow, the appropriate language for the category will be used. If no translation is provided, the original text will be used.'
            );

            getSpecWrapper(wrapper, 'exit-name').forEach((uiNode, idx) =>
                expect(uiNode.text()).toBe(switchProps.node.exits[idx].name)
            );

            getSpecWrapper(wrapper, 'localization-input').forEach((uiNode, idx) => {
                const exitUUID: string = switchProps.node.exits[idx].uuid;
                let value: string = '';
                const localized = localizations.find(
                    localizedObject => localizedObject.getObject().uuid === exitUUID
                );

                if (localized) {
                    if ('name' in localized.localizedKeys) {
                        ({ name: value } = localized.getObject() as Exit);
                    }

                    expect(uiNode.props()).toEqual(
                        expect.objectContaining({
                            'data-spec': 'localization-input',
                            name: exitUUID,
                            placeholder,
                            showLabel: false,
                            value,
                            ComponentMap: switchProps.ComponentMap
                        })
                    );
                }
            });
        });

        it('should render expression form (not translating)', () => {
            /** Note: cases and name field tested in wait_for_expression test above */
            const wrapper: ReactWrapper = mount(
                <SwitchRouterForm
                    {...{
                        ...switchProps,
                        config: getTypeConfig('expression')
                    } as any}
                />,
                { context: switchRouterContext }
            );

            /** Fields */
            expect(
                getSpecWrapper(wrapper, 'lead-in')
                    .find('p')
                    .text()
            ).toBe('If the expression...');

            expect(
                getSpecWrapper(wrapper, 'lead-in')
                    .find('TextInputElement')
                    .props()
            ).toEqual(
                expect.objectContaining({
                    name: 'Expression',
                    showLabel: false,
                    value: '@input',
                    onChange: wrapper.instance().onExpressionChanged,
                    autocomplete: true,
                    required: true,
                    ComponentMap: switchProps.ComponentMap
                })
            );
        });

        it('should render advanced form (translating case args)', () => {
            const wrapper: ReactWrapper = mount(
                <SwitchRouterForm {...{ ...switchPropsTranslating, showAdvanced: true } as any} />,
                {
                    context: switchRouterContext
                }
            );

            expect(getSpecWrapper(wrapper, 'advanced-title').text()).toBe('Rules');

            expect(getSpecWrapper(wrapper, 'advanced-instructions').text()).toBe(
                'Sometimes languages need special rules to route things properly. If a translation is not provided, the original rule will be used.'
            );

            expect(getSpecWrapper(wrapper, 'operator-field').length).toBe(7);

            const localizedArgs = switchProps.node.router.cases.reduce((argsArr, kase) => {
                if (kase.arguments && kase.arguments.length) {
                    const localized = localizations.find(
                        localizedObject => localizedObject.getObject().uuid === kase.uuid
                    );

                    if (localized) {
                        let value: string = '';

                        if ('arguments' in localized.localizedKeys) {
                            const localizedCase = localized.getObject() as Case;

                            if (localizedCase.arguments.length) {
                                [value] = localizedCase.arguments;
                            }
                        }

                        const { verboseName } = switchRouterContext.getOperatorConfig(kase.type);

                        const [argument] = kase.arguments;

                        argsArr.push({
                            value,
                            verboseName,
                            argument,
                            uuid: kase.uuid
                        });
                    }
                }

                return argsArr;
            }, []);

            getSpecWrapper(wrapper, 'operator-field').forEach((uiNode, idx) => {
                const { verboseName, argument, uuid, value } = localizedArgs[idx];

                expect(getSpecWrapper(uiNode, 'verbose-name').text()).toBe(verboseName);
                expect(getSpecWrapper(uiNode, 'argument-to-translate').text()).toBe(argument);
                expect(getSpecWrapper(uiNode, 'translation-input').props()).toEqual({
                    'data-spec': 'translation-input',
                    name: uuid,
                    placeholder,
                    showLabel: false,
                    value,
                    ComponentMap: switchProps.ComponentMap
                });
            });
        });
    });
});
