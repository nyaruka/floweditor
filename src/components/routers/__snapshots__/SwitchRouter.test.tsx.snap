// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwitchRouter > helpers > composeExitMap > should compose a map of exits 1`] = `
Object {
  "63e20770-64c1-4bad-bd73-a34cf6d46866": Object {
    "destination_node_uuid": null,
    "name": "All Responses",
    "uuid": "63e20770-64c1-4bad-bd73-a34cf6d46866",
  },
}
`;

exports[`SwitchRouter > helpers > resolveExits > should resolve exits 1`] = `
Object {
  "cases": Array [
    Object {
      "arguments": Array [
        "red, r",
      ],
      "exit_uuid": "7b245d49-e9e3-4387-b4ad-48deb03528cd",
      "type": "has_any_word",
      "uuid": "87173eee-5270-4233-aede-ca88e14b672a",
    },
  ],
  "defaultExit": "a8bdc1c5-0283-4656-b932-4f4094f4cc7e",
  "exits": Array [
    Object {
      "destination_node_uuid": "e2ecc8de-9774-4b74-a0dc-ca8aea123227",
      "name": "Red",
      "uuid": "7b245d49-e9e3-4387-b4ad-48deb03528cd",
    },
    Object {
      "destination_node_uuid": "533b64e2-5906-4d33-a8e9-64f1cb6c20dd",
      "name": "Other",
      "uuid": "a8bdc1c5-0283-4656-b932-4f4094f4cc7e",
    },
  ],
}
`;

exports[`SwitchRouter > style utils > getItemStyle > should return dragging style when passed a truthy isDragging arg (snapshot) 1`] = `
Object {
  "WebkitTapHighlightColor": "rgba(0,0,0,0)",
  "WebkitTouchCallout": "none",
  "background": "#f2f9fc",
  "borderRadius": 4,
  "boxSizing": "border-box",
  "height": 43,
  "left": 20,
  "margin": 0,
  "opacity": 0.75,
  "outline": "none",
  "pointerEvents": "none",
  "position": "fixed",
  "top": 181.33334350585938,
  "touchAction": "manipulation",
  "transform": "translate(-2px, 17px)",
  "userSelect": "none",
  "width": 590,
  "zIndex": 5000,
}
`;

exports[`SwitchRouter > style utils > getItemStyle > should return not-dragging style when passed a falsy isDragging arg (snapshot) 1`] = `
Object {
  "WebkitTapHighlightColor": "rgba(0,0,0,0)",
  "WebkitTouchCallout": "none",
  "background": false,
  "borderRadius": false,
  "height": false,
  "left": false,
  "opacity": false,
  "outline": "none",
  "pointerEvents": "auto",
  "top": false,
  "touchAction": "manipulation",
  "transform": null,
  "transition": null,
  "userSelect": "none",
  "width": false,
}
`;
